language-plutus-core/src/Language/PlutusCore/Universe/Core.hs

```diff
@@ -0,0 +1,348 @@
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE DefaultSignatures     #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE KindSignatures        #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE TypeOperators         #-}
+{-# LANGUAGE UndecidableInstances  #-}
+{-# LANGUAGE TemplateHaskell       #-}
+
+module Language.PlutusCore.Universe.Core
+    ( Some (..)
+    , In (..)
+    , Of (..)
+    , SomeIn
+    , SomeOf
+    , someValue
+    , Includes (..)
+    , Closed (..)
+    , type (<:)
+    , knownUniOf
+    , GShow (..)
+    , gshow
+    , GEq (..)
+    , (:~:) (..)
+    , GLift (..)
+    ) where
+
+import           Control.DeepSeq
+import           Control.Monad
+import           Data.GADT.Compare
+import           Data.GADT.Show
+import           Data.Hashable
+import           Data.Proxy
+import           Data.Text.Prettyprint.Doc (Pretty (..))
+import           GHC.Exts
+import           Language.Haskell.TH.Lift
+import           Language.Haskell.TH.Syntax
+import           Text.Show.Deriving
+
+{- Note [Universes]
+A universe is a collection of types. It can be finite like
+
+    data U a where
+        UUnit :: U ()
+        UInt  :: U Int
+
+or infinite like
+
+    data U a where
+        UBool :: U Bool
+        UList :: !(U a) -> U [a]
+
+Here are some values of the latter 'U' / the types that they encode
+
+        UBool               / Bool
+        UList UBool         / [Bool]
+        UList (UList UBool) / [[Bool]]
+
+'U' being a GADT allows to package a type from a universe together with a value of that type.
+For example,
+
+    Some (Of UBool True) :: Some (Of U)
+-}
+
+-- | Existential quantification as a data type.
+data Some f = forall a. Some (f a)
+
+-- | A particular type from a universe.
+newtype In uni a = In (uni a)
```

Comment:
> These are nice names, but I think it would be clearer to make them more like their descriptions. I suggest `In` -> `TypeIn`, `SomeIn` -> `SomeTypeIn`, `Of` -> `ValueOf`, `SomeOf` -> `SomeValueOf`

Will do.

> (Maybe also `Of` -> `From`, but I'm not sure about that.)

I use "from" for types: "a type from a universe" and "of" for values: "a value of a type from a universe". I think `Of` -> `From` would be confusing.

Reply:
To me "value from a universe" is more mnemonic than "value of a universe". So I'd have been tempted to say type/in value/from, but I'm not that bothered.

Reply:
That universes trick comes from the dependently typed world (for me at least) as you've noted and there a value of type `*` is a type, so "a value from a universe" still reads like "a type from a universe" for me ('cause types are valid values).

`ValueOf uni` is supposed to expand to "a value of a type from `uni`" (although I do agree that `ValueOf` does not read that way).

Maybe `ValueOfFrom`, I don't know? Or `OfTypeFrom`. Then we can have `TypeFrom` and `OfTypeFrom`, which is kinda nice.

I've already updated to `TypeIn` and `ValueOf`, but I'll reupdate if things sounds good.

Reply:
I think it's fine, the `Type`/`Value` prefixes make it pretty unambiguous now!

If you wanted to go verbose, you'd have to say `ValueOfTypeFrom` which is a bit much :p